File: backend/app/models/goal_journey.py
"""Goal Journey models for the Goals feature."""

from enum import Enum
from datetime import datetime
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field
import uuid


class StepStatus(str, Enum):
    """Status of a journey step."""
    LOCKED = "locked"
    AVAILABLE = "available"
    IN_PROGRESS = "inprogress"
    COMPLETED = "completed"
    SKIPPED = "skipped"
    ALTERNATIVE = "alternative"


class PathType(str, Enum):
    """Type of path in the journey map."""
    MAIN = "main"
    ALTERNATIVE = "alternative"
    COMPLETED = "completed"


class MapPosition(BaseModel):
    """Position of a node on the journey map canvas."""
    x: float = Field(..., ge=0.0, le=1.0, description="X coordinate (0.0 - 1.0)")
    y: float = Field(..., ge=0.0, le=1.0, description="Y coordinate (0.0 - 1.0)")
    layer: int = Field(..., ge=0, description="Depth level in the tree")


class GoalStepBase(BaseModel):
    """Base model for a journey step."""
    title: str = Field(..., min_length=1, description="Step title")
    description: str = Field("", description="What this step entails")
    estimated_days: int = Field(14, ge=1, description="Estimated days to complete")


class GoalStepCreate(GoalStepBase):
    """Model for creating a step (from AI generation)."""
    prerequisites: List[str] = Field(default_factory=list)
    alternatives: List[str] = Field(default_factory=list)
    metadata: Optional[Dict[str, Any]] = None


class GoalStep(GoalStepBase):
    """Full goal step model with all fields."""
    id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    journey_id: str
    custom_title: Optional[str] = None
    order_index: int = 0
    status: StepStatus = StepStatus.LOCKED
    prerequisites: List[str] = Field(default_factory=list)
    alternatives: List[str] = Field(default_factory=list)
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    notes: List[str] = Field(default_factory=list)
    metadata: Optional[Dict[str, Any]] = None
    position: MapPosition = Field(
        default_factory=lambda: MapPosition(x=0.5, y=0.0, layer=0)
    )
    path_type: PathType = PathType.MAIN
    actual_days_spent: Optional[int] = None
    created_at: datetime = Field(default_factory=datetime.utcnow)

    @property
    def display_title(self) -> str:
        """Return custom title if set, otherwise original title."""
        return self.custom_title or self.title

    @property
    def is_unlocked(self) -> bool:
        """Whether the step can be interacted with."""
        return self.status not in (StepStatus.LOCKED, StepStatus.ALTERNATIVE)

    @property
    def is_on_main_path(self) -> bool:
        """Whether this step is on the main journey path."""
        return self.path_type in (PathType.MAIN, PathType.COMPLETED)

    class Config:
        from_attributes = True


class GoalJourneyCreate(BaseModel):
    """Model for generating a new journey."""
    goal_content: str = Field(..., min_length=10, description="The user's goal")
    goal_reason: Optional[str] = Field(None, description="Why this goal is important")
    goal_id: Optional[str] = Field(None, description="Reference to existing goal")


class GoalJourney(BaseModel):
    """Complete goal journey with all steps and metadata."""
    id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    user_id: str
    goal_id: Optional[str] = None
    goal_content: str = Field(..., description="The destination - user's main goal")
    goal_reason: Optional[str] = None
    steps: List[GoalStep] = Field(default_factory=list)
    current_step_index: int = 0
    overall_progress: float = Field(0.0, ge=0.0, le=1.0)
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: Optional[datetime] = None
    journey_started_at: datetime = Field(default_factory=datetime.utcnow)
    is_ai_generated: bool = True
    ai_notes: Optional[str] = None
    map_width: float = 1000.0
    map_height: float = 2000.0

    @property
    def main_path(self) -> List[GoalStep]:
        """Get only the main path steps (excluding alternatives)."""
        return sorted(
            [s for s in self.steps if s.is_on_main_path],
            key=lambda s: s.order_index
        )

    @property
    def current_step(self) -> Optional[GoalStep]:
        """Get the current active step."""
        main = self.main_path
        if 0 <= self.current_step_index < len(main):
            return main[self.current_step_index]
        # Fallback: find in_progress or available step
        for step in main:
            if step.status == StepStatus.IN_PROGRESS:
                return step
        for step in main:
            if step.status == StepStatus.AVAILABLE:
                return step
        return main[0] if main else None

    @property
    def completed_steps(self) -> List[GoalStep]:
        """Get completed steps."""
        return [s for s in self.main_path if s.status == StepStatus.COMPLETED]

    @property
    def remaining_steps(self) -> List[GoalStep]:
        """Get remaining steps (not completed or skipped)."""
        return [
            s for s in self.main_path
            if s.status not in (StepStatus.COMPLETED, StepStatus.SKIPPED)
        ]

    @property
    def steps_to_destination(self) -> int:
        """Steps until destination."""
        return len(self.remaining_steps)

    @property
    def is_complete(self) -> bool:
        """Is journey complete?"""
        return self.overall_progress >= 1.0 or len(self.remaining_steps) == 0

    class Config:
        from_attributes = True


class JourneyAdjustmentRequest(BaseModel):
    """Request model for AI-powered journey adjustment."""
    journey_id: str
    current_activity: str = Field(
        ...,
        min_length=5,
        description="What the user is currently doing"
    )
    additional_context: Optional[str] = None


class JourneyAdjustmentResponse(BaseModel):
    """Response model for journey adjustment."""
    journey: GoalJourney
    changes_made: List[str] = Field(
        default_factory=list,
        description="Description of changes made"
    )
    ai_message: str = Field(..., description="Encouraging message about the adjustment")


class StepStatusUpdate(BaseModel):
    """Request model for updating step status."""
    status: StepStatus
    notes: Optional[str] = Field(None, description="Optional note to add")


class StepTitleUpdate(BaseModel):
    """Request model for updating step title."""
    custom_title: str = Field(..., min_length=1)


class StepNoteAdd(BaseModel):
    """Request model for adding a note to a step."""
    note: str = Field(..., min_length=1)


class JourneyGenerateRequest(BaseModel):
    """Request model for generating a new journey."""
    goal_content: str = Field(..., min_length=10)
    goal_reason: Optional[str] = None
    goal_id: Optional[str] = None
    # Optional context from notification profile
    identity: Optional[str] = None
    challenges: Optional[List[str]] = None


class JourneyGenerateResponse(BaseModel):
    """Response model for journey generation."""
    journey: GoalJourney
    ai_message: str = Field(..., description="AI's overview of the journey")

File: lib/models/goal_journey.dart
/// Goal Journey data models for the Goals feature.
///
/// These models represent the user's journey toward their primary goal,
/// including steps, progress tracking, and the gamified map visualization.

import 'dart:ui' show Offset;

/// Status of a journey step
enum StepStatus {
  locked, // Not yet reachable (grayed out on map)
  available, // Can be started (faded, solid outline)
  inProgress, // Currently working on (glowing, pulsing)
  completed, // Finished (filled with checkmark)
  skipped, // User chose different path (crossed out, visible)
  alternative, // Alternative path not taken (grayed, shows choice)
}

/// Type of path in the journey map
enum PathType {
  main, // The chosen/active path
  alternative, // A visible but unchosen path
  completed, // Already traversed path
}

/// Position of a node on the journey map canvas
class MapPosition {
  final double x; // X coordinate on canvas (0.0 - 1.0 normalized)
  final double y; // Y coordinate on canvas (0.0 - 1.0 normalized)
  final int layer; // Depth level in the tree (0 = start, max = destination)

  const MapPosition({required this.x, required this.y, required this.layer});

  factory MapPosition.fromJson(Map<String, dynamic> json) {
    return MapPosition(
      x: (json['x'] as num).toDouble(),
      y: (json['y'] as num).toDouble(),
      layer: json['layer'] as int,
    );
  }

  Map<String, dynamic> toJson() {
    return {'x': x, 'y': y, 'layer': layer};
  }

  /// Convert to Offset for rendering
  Offset toOffset(double canvasWidth, double canvasHeight) {
    return Offset(x * canvasWidth, y * canvasHeight);
  }
}

/// Estimated Time of Achievement data
class ETAData {
  final DateTime? estimatedCompletionDate;
  final int totalEstimatedDays;
  final int daysElapsed;
  final int stepsCompleted;
  final double averageDaysPerStep;
  final double velocityScore; // > 1.0 = ahead, 1.0 = on track, < 1.0 = behind
  final String displayText;

  const ETAData({
    this.estimatedCompletionDate,
    required this.totalEstimatedDays,
    required this.daysElapsed,
    required this.stepsCompleted,
    required this.averageDaysPerStep,
    required this.velocityScore,
    required this.displayText,
  });

  factory ETAData.initial() => const ETAData(
    estimatedCompletionDate: null,
    totalEstimatedDays: 0,
    daysElapsed: 0,
    stepsCompleted: 0,
    averageDaysPerStep: 14.0,
    velocityScore: 1.0,
    displayText: 'Calculating...',
  );

  factory ETAData.fromJson(Map<String, dynamic> json) {
    return ETAData(
      estimatedCompletionDate: json['estimated_completion_date'] != null
          ? DateTime.parse(json['estimated_completion_date'] as String)
          : null,
      totalEstimatedDays: json['total_estimated_days'] as int,
      daysElapsed: json['days_elapsed'] as int,
      stepsCompleted: json['steps_completed'] as int,
      averageDaysPerStep: (json['average_days_per_step'] as num).toDouble(),
      velocityScore: (json['velocity_score'] as num).toDouble(),
      displayText: json['display_text'] as String,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'estimated_completion_date': estimatedCompletionDate?.toIso8601String(),
      'total_estimated_days': totalEstimatedDays,
      'days_elapsed': daysElapsed,
      'steps_completed': stepsCompleted,
      'average_days_per_step': averageDaysPerStep,
      'velocity_score': velocityScore,
      'display_text': displayText,
    };
  }
}

/// Individual step in the goal journey
class GoalStep {
  final String id;
  final String journeyId;
  final String title;
  final String? customTitle;
  final String description;
  final int order;
  final StepStatus status;
  final List<String> prerequisites;
  final List<String> alternatives;
  final DateTime? startedAt;
  final DateTime? completedAt;
  final List<String> notes;
  final Map<String, dynamic>? metadata;
  final MapPosition position;
  final PathType pathType;
  final int estimatedDays;
  final int? actualDaysSpent;
  final DateTime createdAt;

  const GoalStep({
    required this.id,
    required this.journeyId,
    required this.title,
    this.customTitle,
    required this.description,
    required this.order,
    required this.status,
    this.prerequisites = const [],
    this.alternatives = const [],
    this.startedAt,
    this.completedAt,
    this.notes = const [],
    this.metadata,
    required this.position,
    required this.pathType,
    required this.estimatedDays,
    this.actualDaysSpent,
    required this.createdAt,
  });

  /// Display title (custom if set, otherwise original)
  String get displayTitle => customTitle ?? title;

  /// Whether the step can be interacted with
  bool get isUnlocked =>
      status != StepStatus.locked && status != StepStatus.alternative;

  /// Whether this step is on the main journey path
  bool get isOnMainPath =>
      pathType == PathType.main || pathType == PathType.completed;

  /// Whether this step is the current one
  bool get isCurrent => status == StepStatus.inProgress;

  factory GoalStep.fromJson(Map<String, dynamic> json) {
    return GoalStep(
      id: json['id'] as String,
      journeyId: json['journey_id'] as String,
      title: json['title'] as String,
      customTitle: json['custom_title'] as String?,
      description: json['description'] as String? ?? '',
      order: json['order_index'] as int? ?? json['order'] as int? ?? 0,
      status: StepStatus.values.firstWhere(
        (s) => s.name.toLowerCase() == (json['status'] as String?)?.toLowerCase(),
        orElse: () => StepStatus.locked,
      ),
      prerequisites:
          (json['prerequisites'] as List<dynamic>?)
              ?.map((e) => e as String)
              .toList() ??
          [],
      alternatives:
          (json['alternatives'] as List<dynamic>?)
              ?.map((e) => e as String)
              .toList() ??
          [],
      startedAt: json['started_at'] != null
          ? DateTime.parse(json['started_at'] as String)
          : null,
      completedAt: json['completed_at'] != null
          ? DateTime.parse(json['completed_at'] as String)
          : null,
      notes:
          (json['notes'] as List<dynamic>?)?.map((e) => e as String).toList() ??
          [],
      metadata: json['metadata'] as Map<String, dynamic>?,
      position: json['position'] != null
          ? MapPosition.fromJson(json['position'] as Map<String, dynamic>)
          : MapPosition(
              x: 0.5,
              y: (json['order'] as int? ?? 0) * 0.1,
              layer: json['order'] as int? ?? 0,
            ),
      pathType: PathType.values.firstWhere(
        (p) => p.name == (json['path_type'] as String?)?.toLowerCase(),
        orElse: () => PathType.main,
      ),
      estimatedDays: json['estimated_days'] as int? ?? 14,
      actualDaysSpent: json['actual_days_spent'] as int?,
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'] as String)
          : DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'journey_id': journeyId,
      'title': title,
      'custom_title': customTitle,
      'description': description,
      'order_index': order,
      'status': status.name.toLowerCase(),
      'prerequisites': prerequisites,
      'alternatives': alternatives,
      'started_at': startedAt?.toIso8601String(),
      'completed_at': completedAt?.toIso8601String(),
      'notes': notes,
      'metadata': metadata,
      'position': position.toJson(),
      'path_type': pathType.name,
      'estimated_days': estimatedDays,
      'actual_days_spent': actualDaysSpent,
      'created_at': createdAt.toIso8601String(),
    };
  }

  GoalStep copyWith({
    String? id,
    String? journeyId,
    String? title,
    String? customTitle,
    String? description,
    int? order,
    StepStatus? status,
    List<String>? prerequisites,
    List<String>? alternatives,
    DateTime? startedAt,
    DateTime? completedAt,
    List<String>? notes,
    Map<String, dynamic>? metadata,
    MapPosition? position,
    PathType? pathType,
    int? estimatedDays,
    int? actualDaysSpent,
    DateTime? createdAt,
  }) {
    return GoalStep(
      id: id ?? this.id,
      journeyId: journeyId ?? this.journeyId,
      title: title ?? this.title,
      customTitle: customTitle ?? this.customTitle,
      description: description ?? this.description,
      order: order ?? this.order,
      status: status ?? this.status,
      prerequisites: prerequisites ?? this.prerequisites,
      alternatives: alternatives ?? this.alternatives,
      startedAt: startedAt ?? this.startedAt,
      completedAt: completedAt ?? this.completedAt,
      notes: notes ?? this.notes,
      metadata: metadata ?? this.metadata,
      position: position ?? this.position,
      pathType: pathType ?? this.pathType,
      estimatedDays: estimatedDays ?? this.estimatedDays,
      actualDaysSpent: actualDaysSpent ?? this.actualDaysSpent,
      createdAt: createdAt ?? this.createdAt,
    );
  }

  @override
  String toString() {
    return 'GoalStep(id: $id, title: $displayTitle, status: $status, order: $order)';
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is GoalStep && other.id == id;
  }

  @override
  int get hashCode => id.hashCode;
}

/// Complete goal journey with all steps and metadata
class GoalJourney {
  final String id;
  final String userId;
  final String? goalId;
  final String goalContent; // The destination - user's main goal
  final String? goalReason;
  final List<GoalStep> steps;
  final int currentStepIndex;
  final double overallProgress; // 0.0 to 1.0
  final DateTime createdAt;
  final DateTime? updatedAt;
  final DateTime journeyStartedAt;
  final bool isAIGenerated;
  final String? aiNotes;
  final double mapWidth;
  final double mapHeight;

  const GoalJourney({
    required this.id,
    required this.userId,
    this.goalId,
    required this.goalContent,
    this.goalReason,
    required this.steps,
    required this.currentStepIndex,
    required this.overallProgress,
    required this.createdAt,
    this.updatedAt,
    required this.journeyStartedAt,
    required this.isAIGenerated,
    this.aiNotes,
    this.mapWidth = 1000.0,
    this.mapHeight = 2000.0,
  });

  /// Get only the main path steps (excluding alternatives)
  List<GoalStep> get mainPath =>
      steps.where((s) => s.isOnMainPath).toList()
        ..sort((a, b) => a.order.compareTo(b.order));

  /// Get the current active step
  GoalStep? get currentStep {
    if (currentStepIndex >= 0 && currentStepIndex < mainPath.length) {
      return mainPath[currentStepIndex];
    }
    return mainPath.firstWhere(
      (s) => s.status == StepStatus.inProgress,
      orElse: () => mainPath.firstWhere(
        (s) => s.status == StepStatus.available,
        orElse: () => mainPath.first,
      ),
    );
  }

  /// Get completed steps
  List<GoalStep> get completedSteps =>
      mainPath.where((s) => s.status == StepStatus.completed).toList();

  /// Get remaining steps
  List<GoalStep> get remainingSteps => mainPath
      .where(
        (s) =>
            s.status != StepStatus.completed && s.status != StepStatus.skipped,
      )
      .toList();

  /// Steps until destination
  int get stepsToDestination => remainingSteps.length;

  /// Is journey complete?
  bool get isComplete => overallProgress >= 1.0 || remainingSteps.isEmpty;

  factory GoalJourney.fromJson(Map<String, dynamic> json) {
    final stepsJson = json['steps'] as List<dynamic>? ?? [];
    return GoalJourney(
      id: json['id'] as String,
      userId: json['user_id'] as String,
      goalId: json['goal_id'] as String?,
      goalContent: json['goal_content'] as String,
      goalReason: json['goal_reason'] as String?,
      steps: stepsJson
          .map((s) => GoalStep.fromJson(s as Map<String, dynamic>))
          .toList(),
      currentStepIndex: json['current_step_index'] as int? ?? 0,
      overallProgress: (json['overall_progress'] as num?)?.toDouble() ?? 0.0,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: json['updated_at'] != null
          ? DateTime.parse(json['updated_at'] as String)
          : null,
      journeyStartedAt: json['journey_started_at'] != null
          ? DateTime.parse(json['journey_started_at'] as String)
          : DateTime.parse(json['created_at'] as String),
      isAIGenerated: json['is_ai_generated'] as bool? ?? true,
      aiNotes: json['ai_notes'] as String?,
      mapWidth: (json['map_width'] as num?)?.toDouble() ?? 1000.0,
      mapHeight: (json['map_height'] as num?)?.toDouble() ?? 2000.0,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'user_id': userId,
      'goal_id': goalId,
      'goal_content': goalContent,
      'goal_reason': goalReason,
      'steps': steps.map((s) => s.toJson()).toList(),
      'current_step_index': currentStepIndex,
      'overall_progress': overallProgress,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt?.toIso8601String(),
      'journey_started_at': journeyStartedAt.toIso8601String(),
      'is_ai_generated': isAIGenerated,
      'ai_notes': aiNotes,
      'map_width': mapWidth,
      'map_height': mapHeight,
    };
  }

  GoalJourney copyWith({
    String? id,
    String? userId,
    String? goalId,
    String? goalContent,
    String? goalReason,
    List<GoalStep>? steps,
    int? currentStepIndex,
    double? overallProgress,
    DateTime? createdAt,
    DateTime? updatedAt,
    DateTime? journeyStartedAt,
    bool? isAIGenerated,
    String? aiNotes,
    double? mapWidth,
    double? mapHeight,
  }) {
    return GoalJourney(
      id: id ?? this.id,
      userId: userId ?? this.userId,
      goalId: goalId ?? this.goalId,
      goalContent: goalContent ?? this.goalContent,
      goalReason: goalReason ?? this.goalReason,
      steps: steps ?? this.steps,
      currentStepIndex: currentStepIndex ?? this.currentStepIndex,
      overallProgress: overallProgress ?? this.overallProgress,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      journeyStartedAt: journeyStartedAt ?? this.journeyStartedAt,
      isAIGenerated: isAIGenerated ?? this.isAIGenerated,
      aiNotes: aiNotes ?? this.aiNotes,
      mapWidth: mapWidth ?? this.mapWidth,
      mapHeight: mapHeight ?? this.mapHeight,
    );
  }

  @override
  String toString() {
    return 'GoalJourney(id: $id, goal: $goalContent, progress: ${(overallProgress * 100).toInt()}%, steps: ${steps.length})';
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is GoalJourney && other.id == id;
  }

  @override
  int get hashCode => id.hashCode;
}

/// Request model for journey adjustment
class JourneyAdjustmentRequest {
  final String journeyId;
  final String currentActivity;
  final String? additionalContext;

  const JourneyAdjustmentRequest({
    required this.journeyId,
    required this.currentActivity,
    this.additionalContext,
  });

  Map<String, dynamic> toJson() {
    return {
      'journey_id': journeyId,
      'current_activity': currentActivity,
      'additional_context': additionalContext,
    };
  }
}

/// Response model for journey adjustment
class JourneyAdjustmentResponse {
  final GoalJourney journey;
  final List<String> changesMade;
  final String aiMessage;

  const JourneyAdjustmentResponse({
    required this.journey,
    required this.changesMade,
    required this.aiMessage,
  });

  factory JourneyAdjustmentResponse.fromJson(Map<String, dynamic> json) {
    return JourneyAdjustmentResponse(
      journey: GoalJourney.fromJson(json['journey'] as Map<String, dynamic>),
      changesMade: (json['changes_made'] as List<dynamic>)
          .map((e) => e as String)
          .toList(),
      aiMessage: json['ai_message'] as String,
    );
  }
}
